### PR TITLE
[RLlib] Make sure torch and tf behave the same wrt conv2d nets.

### PR DESCRIPTION
**Pull request by sven1977**
This PR makes sure tf and torch behave the same wrt creating/wrapping Conv2D networks (e.g. for `use_lstm=True + image observations` or `image preprocessor for SAC[TF|Torch]Model`).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/models/torch/misc.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ def initializer(tensor):
     return initializer
 
 
-def valid_padding(in_size, filter_size, stride_size):
+def same_padding(in_size, filter_size, stride_size):
```
Comment (sven1977) [inline rllib/models/torch/misc.py:18]: The old name was VERY confusing :) being "valid_padding", where we wanted to imitate tf's "same" behavior.

File: rllib/models/torch/visionnet.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
+import numpy as np
```
Comment (sven1977) [inline rllib/models/torch/visionnet.py:1]: The changes in this file make sure that torch now behaves exactly the same as tf wrt Conv2D Models, which is:
- if `num_outputs`: end with Conv2D [1,1] layer with num_outputs dim.
- else: end with flatten layer with dim being determined by the last Conv2D's setup.
This allows for simple wrapping with e.g. an LSTM (or any other (custom) Model).

File: rllib/models/tf/visionnet.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import numpy as np
```
Comment (sven1977) [inline rllib/models/tf/visionnet.py:1]: The changes here make sure that Conv2D behave exactly the same as FullyConnected wrt their wrapping behavior.

File: rllib/tuned_examples/ars/cartpole-ars.yaml [hunk 4]
```diff
@@ -3,7 +3,7 @@ cartpole-ars:
     run: ARS
     stop:
         episode_reward_mean: 150
-        timesteps_total: 500000
+        timesteps_total: 1000000
```
Comment (sven1977) [inline rllib/tuned_examples/ars/cartpole-ars.yaml:6]: Safety. ES tends to "stop-out" in this test.

File: rllib/models/tf/visionnet.py [hunk 5]
```diff
@@ -59,12 +63,20 @@ def __init__(self, obs_space, action_space, num_outputs, model_config,
...
+
+            if num_outputs:
+                conv_out = tf.keras.layers.Conv2D(
+                    num_outputs,
+                    [1, 1],
+                    activation=None,
+                    padding="same",
+                    data_format="channels_last",
+                    name="conv_out")(last_layer)
+            else:
```
Comment (sven1977) [inline rllib/models/tf/visionnet.py:75]: If `num_outputs` is None (e.g. when wrapping the Conv2D with a custom or LSTM Model), the last layer will be a convenient flatten one so we can simply connect the wrapper to this Conv2D Model.

self.num_outputs will then be set to the number of nodes in this flatten layer (convenience to the wrapping Model).

File: rllib/BUILD [hunk 6]
```diff
@@ -891,19 +891,6 @@ py_test(
         ]
 )
 
-py_test(
```
Comment (sven1977) [inline rllib/BUILD:894]: these are no longer necessary. They are covered by PPO's compilation tests (test_ppo.py).

File: rllib/models/tf/visionnet.py [hunk 7]
```diff
@@ -35,10 +38,11 @@ def __init__(self, obs_space, action_space, num_outputs, model_config,
                 padding="same",
                 data_format="channels_last",
                 name="conv{}".format(i))(last_layer)
+
         out_size, kernel, stride = filters[-1]
 
         # No final linear: Last layer is a Conv2D and uses num_outputs.
-        if no_final_linear:
+        if no_final_linear and num_outputs:
```
Comment (ericl) [inline rllib/models/tf/visionnet.py:45]: When do we ever set this to None? I don't see it in the modelv2 docs.

Comment (sven1977) [reply rllib/models/tf/visionnet.py:45]: I'll add this to the docs.

Leaving it None (or 0) simply means, we don't care about the `num_outputs`, the Model will figure this out by itself, will not generate a logits layer or last layer and just set its `self.num_outputs` to whatever that'll be (after creating the underlying model).

This has been around for some time in fcnets (see SAC model when not using a Conv2D state preprocessor). Just added it to conv2d nets as well to make them behave the same.
With this in place, we can get rid of the artificially hard-coding num_outputs=256 in case of SAC image space (Conv2D).
```
    if config["use_state_preprocessor"]:
        num_outputs = 256  # Flatten last Conv2D to this many nodes.
    else:
        num_outputs = 0
```

Without this option, when you look at the `make_model` for DQN in dqn_torch/tf_policy, we need to somehow infer the last layer size from the given config (fcnet_hiddens), which is unwieldy (we use 256 hard-coded!). When we create custom models, we usually wouldn't know, what the underlying default model will be (after all, this is the exact reason, we use ModelCatalog to figure this out for us).

**Review by ericl — APPROVED** [review]
